torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -198,21 +208,49 @@ Tensor pow_backward_self(Tensor grad, const Tensor & self, const Tensor & expone
 // d(a^b)/db = 0 for a > 0 and b -> +0.
 // Currently, tensorflow agrees with us.
 Tensor pow_backward_exponent(Tensor grad, const Tensor& self, const Tensor& exponent, Tensor result) {
-  return grad * at::where(at::logical_and(self == 0, exponent >= 0),
+  auto cond = [](auto self, auto exp) {
+      if (exp.is_complex()) {
+        auto is_real_exp = at::logical_and(at::imag(exp) == 0, at::real(exp) >= 0);
+        return at::logical_and(self == 0, is_real_exp);
+      } else {
+        return at::logical_and(self == 0, exp >= 0);
+      }
+    };
+  auto out = grad * at::where(cond(self, exponent),
                           at::zeros({}, grad.options()),
-                          result * self.log());
+                          (result * self.log()).conj());
+  return handle_r_to_c(exponent, out);
 }
 
 Tensor pow_backward_exponent(Tensor grad, const Scalar & base, const Tensor& exponent, Tensor result) {
-  if (base.toDouble() == 0) {
-    return grad * at::where(exponent >= 0,
+  auto base_ = base.isComplex() ? base.toComplexDouble() : base.toDouble();
+  auto grad_lambda = [](auto a, auto b) { return (a * std::log(b)).conj(); };
+  if (base_ == 0.0) {
+    auto cond = [](auto exp) {
+      if (exp.is_complex()) {
+        return at::logical_and(at::imag(exp) == 0, at::real(exp) >= 0);
+      } else {
+        return exp >=0;
+      }
+    };
+    auto out = grad * at::where(cond(exponent),
                             at::zeros({}, grad.options()),
-                            result * std::log(base.toDouble()));
+                            grad_lambda(result, base_));
+    return handle_r_to_c(exponent, out);
   } else {
-    return grad * result * std::log(base.toDouble());
+    auto out = grad * grad_lambda(result, base_);
+    return handle_r_to_c(exponent, out);
   }
 }
 
+Tensor angle_backward(Tensor grad, const Tensor& self) {
+  auto out = grad * self / self.abs().pow(2);
```

Comment:
This looks different that the previous formula (casting of the gradient in particular). Is that expected?

Reply:
The previous formula didn't work correctly with real numbers. Actually on more thought, `torch.angle(real_tensors)` always returns zeros, so we can simply return grad (as can also be seen when the formula is simplified).

Reply:
Also this will return nan at 0. Is that expected?

Reply:
Also this will return nan at 0. Is that expected?

Reply:
good catch. No, I don't think it's correct. It should return zero for the case when `abs = 0`. updated!